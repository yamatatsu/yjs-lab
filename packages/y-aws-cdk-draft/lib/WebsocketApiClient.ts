import { ApiGatewayManagementApi } from "@aws-sdk/client-apigatewaymanagementapi";

export default class WebsocketApiClient {
  private readonly apiGatewayManagementApi: ApiGatewayManagementApi;
  constructor(websocketEndpoint: string) {
    this.apiGatewayManagementApi = new ApiGatewayManagementApi({
      apiVersion: "2018-11-29",
      endpoint: websocketEndpoint,
    });
  }

  async broadcast(
    message: string,
    senderConnectionId: string,
    connectionIds: string[]
  ) {
    await Promise.all(
      connectionIds
        .filter((connectionId) => senderConnectionId !== connectionId)
        .map(async (connectionId) => {
          try {
            await this.send(connectionId, message);
          } catch (error) {
            console.error(error);
          }
        })
    );
  }

  async reply(message: string, senderConnectionId: string) {
    await this.send(senderConnectionId, message);
  }

  private async send(connectionId: string, message: string): Promise<void> {
    await this.apiGatewayManagementApi.postToConnection({
      ConnectionId: connectionId,
      Data: Buffer.from(message),
    });
  }
}
