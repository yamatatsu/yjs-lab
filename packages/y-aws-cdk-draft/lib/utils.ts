import type {
  APIGatewayProxyWebsocketEventV2WithRequestContext,
  APIGatewayEventWebsocketRequestContextV2,
  APIGatewayProxyResultV2,
} from "aws-lambda";

type RequestContext = Pick<
  APIGatewayEventWebsocketRequestContextV2,
  "connectionId" | "domainName" | "stage"
> & { authorizer: { roomId: string } };

type Event = APIGatewayProxyWebsocketEventV2WithRequestContext<RequestContext>;
type Response = Promise<APIGatewayProxyResultV2<never>>;

export type Handler = (event: Event) => Response;

export function getRoomId(event: Event) {
  return event.requestContext.authorizer.roomId;
}
export function getConnectionId(event: Event) {
  return event.requestContext.connectionId;
}
export function getBody(event: Event) {
  return event.body;
}
export function getWebsocketEndpoint(event: Event) {
  const { domainName, stage } = event.requestContext;
  return `https://${domainName}/${stage}`;
}
