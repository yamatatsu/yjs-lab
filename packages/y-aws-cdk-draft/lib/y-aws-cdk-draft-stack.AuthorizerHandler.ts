import type { APIGatewayIAMAuthorizerResult } from "aws-lambda";

type Event = { headers: Record<string, string>; methodArn: string };
type Response = Promise<APIGatewayIAMAuthorizerResult | "Unauthorized">;

export const handler = async (event: Event): Response => {
  console.log(JSON.stringify(event, null, 2));

  const token = event.headers.Authorization;
  const roomId = event.headers["Sec-WebSocket-Protocol"];

  if (isUnauthorized(token)) {
    console.info("Unauthorized");
    return "Unauthorized";
  }

  console.info("Authorized");
  return getPolicy(roomId, event.methodArn);
};

// ==============================
// lib

function isUnauthorized(token?: string): boolean {
  return !token || token !== process.env.WEBSOCKET_TOKEN;
}

function getPolicy(
  roomId: string,
  methodArn: string
): APIGatewayIAMAuthorizerResult {
  return {
    principalId: "me",
    policyDocument: {
      Version: "2012-10-17",
      Statement: [
        {
          Action: "execute-api:Invoke",
          Effect: "Allow",
          Resource: methodArn,
        },
      ],
    },
    context: {
      roomId,
    },
  };
}
