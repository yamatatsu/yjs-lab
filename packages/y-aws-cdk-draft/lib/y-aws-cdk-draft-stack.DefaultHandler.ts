import { ApiGatewayManagementApi } from "@aws-sdk/client-apigatewaymanagementapi";
import { DB } from "./db/dynamodb";
import {
  Handler,
  getConnectionId,
  getRoomId,
  getBody,
  getWebsocketEndpoint,
} from "./utils";

export const handler: Handler = async (event) => {
  console.log(JSON.stringify(event, null, 2));

  const senderConnectionId = getConnectionId(event);
  const roomId = getRoomId(event);
  const websocketEndpoint = getWebsocketEndpoint(event);
  const body = getBody(event) ?? "[empty-body]";

  const items = await DB.query({
    KeyConditionExpression: "pk = :pk",
    ExpressionAttributeValues: { ":pk": roomId },
  });

  const send = apiGatewayManagementApi(websocketEndpoint);

  await Promise.all(
    items
      .filter((item) => senderConnectionId !== item.connectionId)
      .map(async (item) => {
        try {
          await send(item.connectionId, body);
        } catch (error) {
          console.error(error);
        }
      })
  );

  return { statusCode: 200, body: "success" };
};

function apiGatewayManagementApi(websocketEndpoint: string) {
  const apiGateway = new ApiGatewayManagementApi({
    apiVersion: "2018-11-29",
    endpoint: websocketEndpoint,
  });
  return (connectionId: string, body: string) =>
    apiGateway.postToConnection({
      ConnectionId: connectionId,
      Data: Buffer.from(body),
    });
}
