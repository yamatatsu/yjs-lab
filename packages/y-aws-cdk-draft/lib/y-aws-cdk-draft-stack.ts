import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import * as ssm from "aws-cdk-lib/aws-ssm";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as dynamodb from "aws-cdk-lib/aws-dynamodb";
import * as apig from "@aws-cdk/aws-apigatewayv2-alpha";
import { WebSocketLambdaIntegration } from "@aws-cdk/aws-apigatewayv2-integrations-alpha";
import { WebSocketLambdaAuthorizer } from "@aws-cdk/aws-apigatewayv2-authorizers-alpha";

export class WebSocketApiStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const table = new dynamodb.Table(this, "Table", {
      tableName: "y-aws-cdk-table",
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      partitionKey: { name: "pk", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "sk", type: dynamodb.AttributeType.STRING },
    });
    const yDynamodbTable = new dynamodb.Table(this, "YDynamodbTable", {
      tableName: "y-dynamodb",
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      partitionKey: { name: "docName", type: dynamodb.AttributeType.STRING },
      sortKey: { name: "sortKey", type: dynamodb.AttributeType.STRING },
    });

    const authorizerHandler = new NodejsFunction(this, "AuthorizerHandler", {
      entry: "../y-aws-cdk-draft-lambda/src/authorizer-handler.ts",
      runtime: lambda.Runtime.NODEJS_18_X,
      environment: {
        WEBSOCKET_TOKEN: ssm.StringParameter.fromStringParameterName(
          this,
          "WEBSOCKET_TOKEN",
          "/y-aws-cdk/websocket-token"
        ).stringValue,
      },
    });
    const connectHandler = new NodejsFunction(this, "ConnectHandler", {
      entry: "../y-aws-cdk-draft-lambda/src/connect-handler.ts",
      runtime: lambda.Runtime.NODEJS_18_X,
      environment: {
        TABLE_NAME: table.tableName,
      },
    });
    const disconnectHandler = new NodejsFunction(this, "DisconnectHandler", {
      entry: "../y-aws-cdk-draft-lambda/src/disconnect-handler.ts",
      runtime: lambda.Runtime.NODEJS_18_X,
      environment: {
        TABLE_NAME: table.tableName,
      },
    });
    const defaultHandler = new NodejsFunction(this, "DefaultHandler", {
      entry: "../y-aws-cdk-draft-lambda/src/default-handler.ts",
      runtime: lambda.Runtime.NODEJS_18_X,
      environment: {
        TABLE_NAME: table.tableName,
      },
    });

    table.grantWriteData(connectHandler);
    table.grantWriteData(disconnectHandler);
    table.grantReadData(defaultHandler);
    yDynamodbTable.grantReadWriteData(connectHandler);
    yDynamodbTable.grantReadWriteData(defaultHandler);

    const webSocketApi = new apig.WebSocketApi(this, "WebSocketApi", {
      connectRouteOptions: {
        integration: new WebSocketLambdaIntegration(
          "ConnectIntegration",
          connectHandler
        ),
        authorizer: new WebSocketLambdaAuthorizer(
          "Authorizer",
          authorizerHandler,
          // use `Sec-WebSocket-Protocol` header for auth, because `WebSocket` class cannot set `Authorization` request header
          { identitySource: ["route.request.header.Sec-WebSocket-Protocol"] }
        ),
      },
      disconnectRouteOptions: {
        integration: new WebSocketLambdaIntegration(
          "DisconnectIntegration",
          disconnectHandler
        ),
      },
      defaultRouteOptions: {
        integration: new WebSocketLambdaIntegration(
          "DefaultIntegration",
          defaultHandler
        ),
      },
    });

    new apig.WebSocketStage(this, "WebSocketStage", {
      webSocketApi,
      stageName: "dev",
      autoDeploy: true,
    });

    webSocketApi.grantManageConnections(defaultHandler);
  }
}
