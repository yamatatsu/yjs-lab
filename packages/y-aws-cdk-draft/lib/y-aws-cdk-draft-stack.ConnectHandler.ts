import { DB } from "./db/dynamodb";
import { Handler, getConnectionId, getRoomId } from "./utils";

export const handler: Handler = async (event) => {
  console.log(JSON.stringify(event, null, 2));

  const connectionId = getConnectionId(event);
  const roomId = getRoomId(event);

  await DB.put({
    Item: { pk: roomId, sk: connectionId, roomId, connectionId },
  });

  return {
    statusCode: 200,
    body: "Connected",
    headers: { "Sec-WebSocket-Protocol": roomId },
  };
};
