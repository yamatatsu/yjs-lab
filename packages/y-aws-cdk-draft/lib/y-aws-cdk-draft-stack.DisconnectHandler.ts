import { DB } from "./db/dynamodb";
import { Handler, getConnectionId, getRoomId } from "./utils";

export const handler: Handler = async (event) => {
  console.log(JSON.stringify(event, null, 2));

  const connectionId = getConnectionId(event);
  const roomId = getRoomId(event);

  await DB.delete({ Key: { pk: roomId, sk: connectionId } });

  return { statusCode: 200, body: "Disconnected" };
};
