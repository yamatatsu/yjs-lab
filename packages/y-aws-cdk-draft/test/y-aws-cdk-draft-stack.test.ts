import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as lambda from "aws-cdk-lib/aws-lambda";
import { WebSocketApiStack } from "../lib/y-aws-cdk-draft-stack";

test("Snapshot test", () => {
  const app = new cdk.App();
  const stack = new WebSocketApiStack(app, "Target");

  cdk.Aspects.of(stack).add({
    visit(construct) {
      if (
        construct instanceof lambda.CfnFunction &&
        "s3Key" in construct.code
      ) {
        construct.code = { s3Key: "dummy.zip" };
      }
    },
  });

  const template = Template.fromStack(stack);
  expect(template.toJSON()).toMatchSnapshot();
});
