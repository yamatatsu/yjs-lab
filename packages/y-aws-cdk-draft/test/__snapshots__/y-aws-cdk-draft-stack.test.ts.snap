// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "WEBSOCKETTOKENParameter": {
      "Default": "/y-aws-cdk/websocket-token",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AuthorizerHandler0112B303": {
      "DependsOn": [
        "AuthorizerHandlerServiceRole5F40A014",
      ],
      "Properties": {
        "Code": {
          "S3Key": "dummy.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "WEBSOCKET_TOKEN": {
              "Ref": "WEBSOCKETTOKENParameter",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AuthorizerHandlerServiceRole5F40A014",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "AuthorizerHandlerServiceRole5F40A014": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ConnectHandler2FFD52D8": {
      "DependsOn": [
        "ConnectHandlerServiceRoleDefaultPolicy7DE94863",
        "ConnectHandlerServiceRole7E4A9B1F",
      ],
      "Properties": {
        "Code": {
          "S3Key": "dummy.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TABLE_NAME": {
              "Ref": "TableCD117FA1",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "ConnectHandlerServiceRole7E4A9B1F",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ConnectHandlerServiceRole7E4A9B1F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ConnectHandlerServiceRoleDefaultPolicy7DE94863": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TableCD117FA1",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "ConnectHandlerServiceRoleDefaultPolicy7DE94863",
        "Roles": [
          {
            "Ref": "ConnectHandlerServiceRole7E4A9B1F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DefaultHandler604DF7AC": {
      "DependsOn": [
        "DefaultHandlerServiceRoleDefaultPolicy2F57C32F",
        "DefaultHandlerServiceRoleDF00569C",
      ],
      "Properties": {
        "Code": {
          "S3Key": "dummy.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TABLE_NAME": {
              "Ref": "TableCD117FA1",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "DefaultHandlerServiceRoleDF00569C",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "DefaultHandlerServiceRoleDF00569C": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "DefaultHandlerServiceRoleDefaultPolicy2F57C32F": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TableCD117FA1",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": "execute-api:ManageConnections",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":execute-api:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":",
                    {
                      "Ref": "WebSocketApi34BCF99B",
                    },
                    "/*/*/@connections/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DefaultHandlerServiceRoleDefaultPolicy2F57C32F",
        "Roles": [
          {
            "Ref": "DefaultHandlerServiceRoleDF00569C",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DisconnectHandlerCB7ED6F7": {
      "DependsOn": [
        "DisconnectHandlerServiceRoleDefaultPolicy1800B9E5",
        "DisconnectHandlerServiceRoleE54F14F9",
      ],
      "Properties": {
        "Code": {
          "S3Key": "dummy.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "TABLE_NAME": {
              "Ref": "TableCD117FA1",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "DisconnectHandlerServiceRoleE54F14F9",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "DisconnectHandlerServiceRoleDefaultPolicy1800B9E5": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "TableCD117FA1",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "DisconnectHandlerServiceRoleDefaultPolicy1800B9E5",
        "Roles": [
          {
            "Ref": "DisconnectHandlerServiceRoleE54F14F9",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "DisconnectHandlerServiceRoleE54F14F9": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "TableCD117FA1": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
          {
            "AttributeName": "sk",
            "AttributeType": "S",
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "sk",
            "KeyType": "RANGE",
          },
        ],
        "ProvisionedThroughput": {
          "ReadCapacityUnits": 5,
          "WriteCapacityUnits": 5,
        },
        "TableName": "y-aws-cdk-table",
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "WebSocketApi34BCF99B": {
      "Properties": {
        "Name": "WebSocketApi",
        "ProtocolType": "WEBSOCKET",
        "RouteSelectionExpression": "$request.body.action",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "WebSocketApiAuthorizer4A63F12E": {
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B",
        },
        "AuthorizerType": "REQUEST",
        "AuthorizerUri": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":apigateway:",
              {
                "Ref": "AWS::Region",
              },
              ":lambda:path/2015-03-31/functions/",
              {
                "Fn::GetAtt": [
                  "AuthorizerHandler0112B303",
                  "Arn",
                ],
              },
              "/invocations",
            ],
          ],
        },
        "IdentitySource": [
          "route.request.header.Authorization",
        ],
        "Name": "Authorizer",
      },
      "Type": "AWS::ApiGatewayV2::Authorizer",
    },
    "WebSocketApiTargetWebSocketApiAuthorizer65CB8C43PermissionAC3AE8E4": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "AuthorizerHandler0112B303",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "WebSocketApi34BCF99B",
              },
              "/authorizers/",
              {
                "Ref": "WebSocketApiAuthorizer4A63F12E",
              },
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebSocketApiconnectRoute846149DD": {
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B",
        },
        "AuthorizationType": "CUSTOM",
        "AuthorizerId": {
          "Ref": "WebSocketApiAuthorizer4A63F12E",
        },
        "RouteKey": "$connect",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "WebSocketApiconnectRouteConnectIntegration7F1E0FDE",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "WebSocketApiconnectRouteConnectIntegration7F1E0FDE": {
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":apigateway:",
              {
                "Ref": "AWS::Region",
              },
              ":lambda:path/2015-03-31/functions/",
              {
                "Fn::GetAtt": [
                  "ConnectHandler2FFD52D8",
                  "Arn",
                ],
              },
              "/invocations",
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "WebSocketApiconnectRouteConnectIntegrationPermission39398969": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "ConnectHandler2FFD52D8",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "WebSocketApi34BCF99B",
              },
              "/*/*$connect",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebSocketApidefaultRoute05EFD2FA": {
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "$default",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "WebSocketApidefaultRouteDefaultIntegration0865A209",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "WebSocketApidefaultRouteDefaultIntegration0865A209": {
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":apigateway:",
              {
                "Ref": "AWS::Region",
              },
              ":lambda:path/2015-03-31/functions/",
              {
                "Fn::GetAtt": [
                  "DefaultHandler604DF7AC",
                  "Arn",
                ],
              },
              "/invocations",
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "WebSocketApidefaultRouteDefaultIntegrationPermission0418F15D": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "DefaultHandler604DF7AC",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "WebSocketApi34BCF99B",
              },
              "/*/*$default",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebSocketApidisconnectRouteC181A19C": {
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "$disconnect",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "WebSocketApidisconnectRouteDisconnectIntegration94C91381",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "WebSocketApidisconnectRouteDisconnectIntegration94C91381": {
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":apigateway:",
              {
                "Ref": "AWS::Region",
              },
              ":lambda:path/2015-03-31/functions/",
              {
                "Fn::GetAtt": [
                  "DisconnectHandlerCB7ED6F7",
                  "Arn",
                ],
              },
              "/invocations",
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "WebSocketApidisconnectRouteDisconnectIntegrationPermissionAE705904": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "DisconnectHandlerCB7ED6F7",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "WebSocketApi34BCF99B",
              },
              "/*/*$disconnect",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebSocketApirgaerihgaerghlaiehgraruiehgaeluirhgRouteACBBC266": {
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "rgaerihgaerghlaiehgraruiehgaeluirhg",
        "Target": {
          "Fn::Join": [
            "",
            [
              "integrations/",
              {
                "Ref": "WebSocketApirgaerihgaerghlaiehgraruiehgaeluirhgRouteSendMessageIntegrationC7AF680C",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "WebSocketApirgaerihgaerghlaiehgraruiehgaeluirhgRouteSendMessageIntegrationC7AF680C": {
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":apigateway:",
              {
                "Ref": "AWS::Region",
              },
              ":lambda:path/2015-03-31/functions/",
              {
                "Fn::GetAtt": [
                  "DefaultHandler604DF7AC",
                  "Arn",
                ],
              },
              "/invocations",
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "WebSocketApirgaerihgaerghlaiehgraruiehgaeluirhgRouteSendMessageIntegrationPermission7ADC8DB1": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "DefaultHandler604DF7AC",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "WebSocketApi34BCF99B",
              },
              "/*/*rgaerihgaerghlaiehgraruiehgaeluirhg",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebSocketStageC46B7E43": {
      "Properties": {
        "ApiId": {
          "Ref": "WebSocketApi34BCF99B",
        },
        "AutoDeploy": true,
        "StageName": "dev",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
